arangod/IResearch/IResearchLink.cpp

```diff
@@ -1459,61 +1397,48 @@ arangodb::Result IResearchLink::insert(
 #endif
 
   if (!ctx) {
-    SCOPED_LOCK_NAMED(_asyncSelf->mutex(),
-                      lock);  // '_dataStore' can be asynchronously modified
+    // '_dataStore' can be asynchronously modified
+    SCOPED_LOCK_NAMED(_asyncSelf->mutex(), lock);
 
     if (!*_asyncSelf) {
-      return arangodb::Result(
-          TRI_ERROR_ARANGO_INDEX_HANDLE_BAD,  // the current link is no longer
-                                              // valid (checked after ReadLock
-                                              // aquisition)
-          std::string("failed to lock arangosearch link while inserting a "
-                      "document into arangosearch link '") +
-              std::to_string(id()) + "'");
+      // the current link is no longer valid (checked after ReadLock aquisition)
+      return {
+        TRI_ERROR_ARANGO_INDEX_HANDLE_BAD,
+        "failed to lock arangosearch link while inserting a "
+        "document into arangosearch link '" + std::to_string(id()) + "'"
+      };
     }
 
     TRI_ASSERT(_dataStore); // must be valid if _asyncSelf->get() is valid
 
-    // optimization for single-document insert-only transactions
-    if (trx.isSingleOperationTransaction() // only for single-docuemnt transactions
-        && RecoveryState::DONE == _dataStore._recovery) {
-      auto ctx = _dataStore._writer->documents();
+//FIXME try to preserve optimization
```

Comment:
what about this comment? does it mean the optimization is turned off?

Reply:
yes, just a reminder, currently we don't operation transaction tick in advance